File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        map = {"(" : ")", "{" : "}", "[" : "]"}
        stack = []

        for char in s:
            if char in map:
                stack.append(char)
            else:
                if not stack:
                    return False
                
                temp = stack.pop()
                if char != map[temp]:
                    return False
        
        return len(stack) == 0

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char curr = s.charAt(i);
            if (curr == '(') {
                st.push(')');
            } else if (curr == '{') {
                st.push('}');
            } else if (curr == '[') {
                st.push(']');
            } else if (st.isEmpty() || st.pop() != s.charAt(i)) {
                return false;
            } 
        }
        // if still chars left that means every char did not have a match
        return st.isEmpty();
    }
}